File: client/src/models/Talent.ts
import {Option} from "../components/common/InputSelectField";
import Modifier from "./common/Modifier";
import Cost from "./common/Cost";
import Limit from "./common/Limit";
import Skill from "./actor/Skill";
import {ActorSkill} from "./actor/Actor";
import {Difficulty} from "./common/Difficulty";

export default interface Talent {
    id: string
    name: string
    ranked: boolean
    activation: Activation
    tier: Tier
    summary: string
    description: string
    cost: Cost
    limit: Limit
    talentSkills: TalentSkills
    talentStats: TalentStats
    talentSkillCheck: TalentSkillCheck
    talentRollModifiers: TalentRollModifiers[]
    modifiers: Modifier[]
}

export interface TalentSkills {
    potentialCareerSkills: Skill[]
}

export interface TalentStats {
    wounds: number
    strain: number
    soak: number
    defense: number
}

export interface TalentSkillCheck {
    skill: ActorSkill
    difficulty: Difficulty
    opposedSkill: ActorSkill
}

export interface TalentRollModifiers {
    skill: ActorSkill
}

export enum Activation {
    Passive = 'Passive',
    ActiveAction = 'Active (Action)',
    ActiveManeuver = 'Active (Maneuver)',
    ActiveIncidental = 'Active (Incidental)',
    ActiveIncidentalOutOfTurn = 'Active (Incidental, Out of Turn)'
}

export const getActivationOptions = (): Option[] => {
    return Object.values(Activation).map((value) => ({value}))
}

export enum Tier {
    First = 'First',
    Second = 'Second',
    Third = 'Third',
    Fourth = 'Fourth',
    Fifth = 'Fifth'
}

export interface ActorTalent extends Talent {
    ranks: number
}

File: server/src/main/java/com/github/genraven/genesys/domain/talent/Talent.java
package com.github.genraven.genesys.domain.talent;

import com.fasterxml.jackson.annotation.JsonValue;
import com.github.genraven.genesys.domain.Activation;
import com.github.genraven.genesys.domain.Cost;
import com.github.genraven.genesys.domain.Limit;
import com.github.genraven.genesys.domain.modifier.Modifier;
import lombok.AllArgsConstructor;
import lombok.Data;
import lombok.Getter;
import org.springframework.data.annotation.Id;
import org.springframework.data.mongodb.core.mapping.Document;

import java.util.ArrayList;
import java.util.List;

@Data
@Document(collection = "talents")
public class Talent {

    public Talent(final String name) {
        this.name = name;
    }

    protected Talent() {
    }

    @Id
    private String id;
    private String name;
    private Activation activation;
    private Tier tier;
    private boolean ranked = false;
    private String summary;
    private String description;
    private Cost cost = new Cost();
    private Limit limit = new Limit();
    private TalentSkills talentSkills = new TalentSkills();
    private List<Modifier> modifiers = new ArrayList<>();

    @AllArgsConstructor
    @Getter
    public enum Tier {
        FIRST("First"),
        SECOND("Second"),
        THIRD("Third"),
        FOURTH("Fourth"),
        FIFTH("Fifth");

        @JsonValue
        private final String label;
    }
}
